File: src/main/java/algo/depthFirstSearch/IsGraphBipartite_785.java
package algo.depthFirstSearch;

public class IsGraphBipartite_785 {
    public static boolean isBipartite(int[][] graph) {
        var vNum = graph.length;
        var color = new int[vNum];

        for (var i = 0; i < vNum; i++) {
            if (color[i] == 0) {
                if (!colorize(graph, color, -1, i)) {
                    return false;
                }
            }
        }

        return true;
    }

    private static boolean colorize(int[][] graph, int[] color, int from, int to) {
        if (color[to] == 0) {
            if (from != -1) {
                color[to] = color[from] == 1 ? 2 : 1;
            } else {
                // root
                color[to] = 1;
            }
        } else {
            if (from != -1 && color[to] == color[from]) {
                return false;
            }
            return true;
        }
        for (var v : graph[to]) {
            if (!colorize(graph, color, to, v)) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        System.out.println(isBipartite(new int[][]{{1,2,3},{0,2},{1,3},{0,2}}));
        System.out.println(isBipartite(new int[][]{{1,3},{0,2},{1,3},{0,2}}));
        System.out.println(isBipartite(new int[][]{{3},{2,4},{1},{0,4},{1,3}}));
    }
}

File: src/main/kotlin/algo/depthFirstSearch/IsGraphBipartite_785.kt
package algo.depthFirstSearch

import java.util.LinkedList

class KtIsGraphBipartite_785 {
    companion object {
        const val BLACK = -1
        const val RED = 1
        const val NONE = 0
    }
    fun isBipartite(graph: Array<IntArray>): Boolean {
        fun bfs(graph: Array<IntArray>, node: Int, color: IntArray): Boolean {
            val q = LinkedList<Int>()
            q.offer(node)
            color[node] = BLACK
            while (q.isNotEmpty()) {
                val size = q.size
                for (i in 1..size) {
                    val n = q.poll()
                    graph[n].forEach {
                        when (color[it]) {
                            color[n] -> return false
                            NONE -> {
                                color[it] = -1 * color[n]
                                q.offer(it)
                            }
                        }
                    }
                }
            }
            return true
        }

        val color = IntArray(graph.size)
        graph.indices.forEach {
            if (color[it] == NONE && !bfs(graph, it, color)) {
                return false
            }
        }
        return true
    }
}

fun main() {
    val bipartite = KtIsGraphBipartite_785()
    println(bipartite.isBipartite(arrayOf(intArrayOf(1,2,3), intArrayOf(0,2), intArrayOf(0,1,3), intArrayOf(0,2))))
    println(bipartite.isBipartite(arrayOf(intArrayOf(1,3), intArrayOf(0,2), intArrayOf(1,3), intArrayOf(0,2))))
}